test/unit/expression_test.rb

```diff
@@ -0,0 +1,109 @@
+require 'test_helper'
+
+class ExpressionTest < MiniTest::Test
+  def test_literals
+    assert_equal true, Liquid::C::Expression.strict_parse('true')
+    assert_equal false, Liquid::C::Expression.strict_parse('false')
+    assert_equal nil, Liquid::C::Expression.strict_parse('nil')
+    assert_equal nil, Liquid::C::Expression.strict_parse('null')
+
+    empty = Liquid::C::Expression.strict_parse('empty')
+    assert_equal '', empty
+    assert_same empty, Liquid::C::Expression.strict_parse('blank')
+  end
+
+  def test_byte_int
+    assert_equal 127, Liquid::C::Expression.strict_parse('127')
+    assert_equal -128, Liquid::C::Expression.strict_parse('-128')
+  end
+
+  def test_short_int
+    assert_equal 128, Liquid::C::Expression.strict_parse('128')
+    assert_equal -129, Liquid::C::Expression.strict_parse('-129')
+    assert_equal 32767, Liquid::C::Expression.strict_parse('32767')
+    assert_equal -32768, Liquid::C::Expression.strict_parse('-32768')
+  end
+
+  def test_float
+    assert_equal 123.4, Liquid::C::Expression.strict_parse('123.4')
+  end
+
+  def test_string
+    assert_equal "hello", Liquid::C::Expression.strict_parse('"hello"')
+    assert_equal "world", Liquid::C::Expression.strict_parse("'world'")
+  end
+
+  def test_find_static_variable
+    context = Liquid::Context.new({"x" => 123})
+    expr = Liquid::C::Expression.strict_parse('x')
+
+    assert_instance_of(Liquid::C::Expression, expr)
+    assert_equal 123, context.evaluate(expr)
+  end
+
+  def test_find_dynamic_variable
+    context = Liquid::Context.new({"x" => "y", "y" => 42})
+    expr = Liquid::C::Expression.strict_parse('[x]')
```

Comment:
I thought `[x]` on its own was supposed to be a syntax error, and that it required an identifier in front. It looks like this will make `{{ [x] }}` possible, which is currently disallowed

```rb
Liquid::Template.parse("{{ [x] }}", "x" => "y", "y" => 42, error_mode: :strict).render
...
# => Liquid::SyntaxError (Liquid syntax error: [:open_square, "["] is not a valid expression in "{{ [x] }}")
```

I'm not necessarily opposed to making it allowed, but we should get consensus from `@Shopify/guardians-of-the-liquid` and modify this in Ruby too.

Reply:
Oh, I was looking at what was possible with `Liquid::Expression.parse("[x]")`.  I didn't realize we considered a dynamic variable lookup to be a strict parse error.  I change it back to being a strict parse error, since I didn't intend to change it here.

Reply:
I’ll also note that if you try to parse and render it in lax mode, it returns an empty string.

Reply:
```
~/src/liquid(master)$ bundle exec irb -Ilib -rliquid
irb> Liquid::Template.parse('{{ [x] }}').render({ 'x' => 'y', 'y' => 42 })
=> "42"
irb> context = Liquid::Context.new({ 'x' => 'y', 'y' => 42 })
irb> expression = Liquid::Expression.parse('[x]')
irb> context.evaluate(expression)
=> 42
```

Reply:
Oh, it was returning an empty string for you since you were passing the variable values as an argument to parse, when they are an argument to render

Reply:
Ah, obviously 🤦

Reply:
Actually, it looks like this has been the behaviour strict parse behaviour for variables in liquid-c for a long time now.  I think since variable parsing was added to liquid-c (see how `TOKEN_OPEN_SQUARE` is allowed at the start of `parse_variable` in the PR that introduced the function https://github.com/Shopify/liquid-c/pull/13/files#diff-307861feddcb5da5f0af73741d9fdea2R93)

So I'm hesitant to change it.  Especially since it could be a breaking change for strictly parsed code.  I'm also unsure if it was really the intention of the strict parser to remove any features, so we could consider whether this should be considered a strict parser bug in the liquid gem.

Reply:
I agree, in the end it seems we should add support for this to the Ruby strict parser.